File: 2381_-_Shifting_Letters_II/2381.py
class List(list):
    pass

class Solution:
    def shiftingLetters(self, s: str, shifts: list[list[int]]) -> str:
        strlen = len(s)
        s = list(s)
        shift_list = [0] * (strlen + 1)  # Use strlen + 1 to handle boundaries
        FORWARD = 1

        for start, end, direction in shifts:
            delta = 1 if direction == FORWARD else -1
            shift_list[start] += delta
            shift_list[end + 1] -= delta

        for i in range(1, strlen):
            shift_list[i] += shift_list[i - 1]

        for i in range(strlen):
            shift = shift_list[i] % 26
            s[i] = chr((ord(s[i]) - ord('a') + shift) % 26 + ord('a'))

        return ''.join(s)


s = "abc"
shifts = [[0,1,0],[1,2,1],[0,2,1]]

print(Solution().shiftingLetters(s, shifts))

File: 2381_-_Shifting_Letters_II/2381.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int strlen = s.size();
        vector<int> shift_list(strlen + 1, 0);
        const int FORWARD = 1;
        int delta = 0;

        for (auto& shift : shifts) {
            delta = shift[2] == FORWARD ? 1 : -1;
            shift_list[shift[0]] += delta;
            shift_list[shift[1] + 1] -= delta;
        }

        for (int i = 1; i < strlen; i++) {
            shift_list[i] += shift_list[i - 1];
        }

        for (int i = 0; i < strlen; i++) {
            int shift = (shift_list[i] + 26) % 26;
            s[i] = (s[i] - 'a' + shift) % 26 + 'a';
        }

        return s;
    }
};


